crates/liblovely/src/lib.rs

```diff
@@ -0,0 +1,70 @@
+use lovely_core::log::*;
+use lovely_core::sys::{LuaLib, LuaState};
+use std::panic;
+
+use lovely_core::Lovely;
+use once_cell::sync::OnceCell;
+
+static RUNTIME: OnceCell<Lovely> = OnceCell::new();
+
+type LoadBufferX =
+    unsafe extern "C" fn(*mut LuaState, *const u8, isize, *const u8, *const u8) -> u32;
+
+static RECALL: OnceCell<LoadBufferX> = OnceCell::new();
+
+#[no_mangle]
+unsafe extern "C" fn lovely_init(
+    loadbufferx: LoadBufferX,
+    lua_call: *const std::ffi::c_void,
+    lua_pcall: *const std::ffi::c_void,
+    lua_getfield: *const std::ffi::c_void,
+    lua_setfield: *const std::ffi::c_void,
+    lua_gettop: *const std::ffi::c_void,
+    lua_settop: *const std::ffi::c_void,
+    lua_pushvalue: *const std::ffi::c_void,
+    lua_pushcclosure: *const std::ffi::c_void,
+    lua_tolstring: *const std::ffi::c_void,
+) {
+    if RUNTIME.get().is_none() {
+        panic::set_hook(Box::new(|x| {
+            let message = format!("lovely-injector has crashed: \n{x}");
+            error!("{message}");
+        }));
+
+        RECALL.set(loadbufferx).expect("Shit's erroring");
+
+
+        let lua = LuaLib {
+            lua_call: std::mem::transmute(lua_call),
```

Comment:
TODO: do I need to verify these are valid?

Reply:
You could in theory check the first n bytes of each address to assert that it's a valid function prototype, but otherwise there's not much that you can do.

Reply:
I could make sure they aren't null pointers

Reply:
Ah yeah, other than that.

Reply:
I think it's safe to assume I'm not being fed garbage pointers

Reply:
Null checking is the correct thing to do but it's really not that important for this.